web/docs/operators/transformations/yara.md

```diff
@@ -0,0 +1,141 @@
+# yara
+
+Executes Yara rules on byte streams.
+
+## Synopsis
+
+```
+yara [-C|--compiled-rules] [-f|--fast-scan] <rule> [<rule>..]
+```
+
+## Description
+
+The `yara` operator applies [Yara](https://virustotal.github.io/yara/) rules to
+an input of bytes, emitting rule context upon a match.
+
+![Yara Operator](yara-operator.excalidraw.svg)
+
+We modeled the operator after the official [`yara` command-line
+utility](https://yara.readthedocs.io/en/stable/commandline.html) to enable a
+familiar experience for the command users. Similar to the official `yara`
+command, the operator compiles the rules by default, unless you provide the
+option `-C,--compiled-rules`. To quote from the above link:
+
+> This is a security measure to prevent users from inadvertently using compiled
+> rules coming from a third-party. Using compiled rules from untrusted sources
+> can lead to the execution of malicious code in your computer.
+
+### `-C|--compiled-rules`
+
+Interpret the rules as compiled.
+
+When providing this flag, you must exactly provide one rule path as positional
+argument.
+
+### `-f|--fast-scan`
+
+Enable fast matching mode.
+
+### `<rule>`
+
+The path to the Yara rule(s).
+
+If the path is a directory, the operator attempts to recursively add all
+contained files as Yara rules.
+
+## Examples
+
+Scan a file with a set of Yara rules:
+
+```
+load file --mmap evil | yara rule.yara
+```
+
+:::caution
+Note the `--mmap` flag: we need it to deliver a single block of data to the
+`yara` operator. Without `--mmap`, the [`file`](../../connectors/file.md) loader
+will generate a stream of byte chunks and feed them incrementally to `yara`,
+attempting to match `rule.yara` individualy on *every chunk*. This will cause
+false negatives when rule matches would span chunk boundaries.
```

Comment:
It may be better to handle this like the `feather` and `parquet` readers do: Merge the chunks and only then match on them. Warn if you receive multiple.

Reply:
Okay, I think that would require a bit of restructuring but it's possible. How would we do this in streaming mode then? With an outer `group` operator?

Reply:
Yes, you'd need some nested pipeline that actually terminates. With some hypothetical operators:

```
ls path/to/dir
| eval {
  load {{file}}
  | yara
}
| …

Reply:
Another advantage of doing this: operators like `shell` and connectors like `s3` will magically work with `yara` as well, which I don't believe upstream Yara can do. That's a pretty cool thing.

Reply:
True, but we also don't know yet how many users have the need to match YARA rules on large S3 buckets. I think it's not uncommon, but I do lack data.

Reply:
In any case, it's better to have a default that just works and does not depend on some internal batching parameters that we may want to change without breaking a user's setup.

Reply:
> Merge the chunks and only then match on them. Warn if you receive multiple.

I don't understand the second sentence: isn't the idea to buffer all chunks and then match on "one giant" one?

FWIW, the YARA C API has mode where you can give a fixed set of memory blocks before doing a scan. So simply buffering all chunks might actually work. Then we'd execute the scan once we're out of chunks.

Reply:
> > Merge the chunks and only then match on them. Warn if you receive multiple.
> 
> I don't understand the second sentence: isn't the idea to buffer all chunks and then match on "one giant" one?

The idea is to emit a warning about this buffering behavior if we receive multiple chunks. You would still continue and execute the rules on the combined content. Here is the warning that we emit for `feather`:

~~~
// TODO: Loading everything into memory here is far from ideal. We should
// instead load passive stores incrementally. For now we at least warn the
// user that this is experimental.
ctrl.warn(caf::make_error(
    ec::unspecified, fmt::format("the experimental {} parser does "
                                 "not currently load files "
                                 "incrementally and may use an "
                                 "excessive amount of memory; consider "
                                 "using 'from file --mmap'",
                                 name)));
~~~

But that is more of a temporary thing, as we eventually want to have an incremental version of this. I'm not sure if it's applicable here. Unless YARA offers a way to incrementally match and notify us when it's done with a chunk?

> FWIW, the YARA C API has mode where you can give a fixed set of memory blocks before doing a scan. So simply buffering all chunks might actually work. Then we'd execute the scan once we're out of chunks.

That would be nice indeed.

Reply:
> The idea is to emit a warning about this buffering behavior if we receive multiple chunks. You would still continue and execute the rules on the combined content.

I'm struggling with this idea, because it breaks composability. Say I have one file to-be-scanned via Kafka. That's a continuously running pipeline. I *want* to match on every chunk. I think opt-in is fine for this, and that accumulating should still be the default.

> Unless YARA offers a way to incrementally match and notify us when it's done with a chunk?

I'm investigating this right now.

Reply:
> I want to match on every chunk [individually].

We can absolutely do this. However, it would imply that, semantically, we no longer deal with bytestreams, but a rather a stream of byte chunks. That's in and of itself fine, and we also seem to be going in the direction with the introduction of the content type. But this is not the mental model we are currently communicating, and it could create issues in the future. Perhaps it would be wiser to separate the "here are multiple blobs" content from the "here is a stream of bytes".

Reply:
That's a good observation. The streaming semantics are indeed more about chunks.

For this operator, I am going to provide two usage modes:
1. `... | yara <rule>` buffers merges all chunks in a single one, performing one `memcpy` into an accumulating `chunk` with every new chunk arriving
2. `... | yara --incremental <rule>` doesn't do any buffering but applies `rule` to every chunk individually.

Users would do (2) only if they know that they get a stream of semantically well-defined chunks where chunk overlap wouldn't exist by design (e.g., one file per Kafka message).

Reply:
I think this is a good approach for now. (2) might be broken if we would internally rebatch. With a separate type, you would not have to manually specify anything by the way, as this could be automatically detected (and you could therefore not forget it in the Kafka case). But let's stick to your solution for now.

Reply:
I've now switched to a `--blockwise` flag to *opt out* of incremental processing. This is what @thomaspatzke recommended, as it avoids footguns.

The good thing is that I was able to use YARA's scanner API that supports cross-chunk matching!

Regarding `--blockwise`: I needed to find a name that doesn't conflict with the YARA CLI, and since "blocks" are the way YARA users think of memory chunks, I opted for that terminology. Also, `-B` as shorthand is not in conflict with any of the `yara` CLI options.

I'm resolving this thread. You may want to review the `memory_block_vector` class separately that implements the incremental matching.